### PR TITLE
[infra] Add Ubuntu build for Travis CI

### PR DESCRIPTION
**Pull request by archibate**
Related issue comment id = https://github.com/taichi-dev/taichi/pull/651#issuecomment-603372369

[[Click here for the format server]](http://kun.csail.mit.edu:31415/)

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -4,18 +4,40 @@ language: cpp
...
       - MATRIX_EVAL="CC=clang && CXX=clang++ && PYTHON_VERSION=3.7.1 && PYTHON=python"
-  - os: osx
+  - name: "OS X + Python 3.8"
+    if: commit_message =~ /^\[release\]/
+    os: osx
     osx_image: xcode10.3
     env:
       - MATRIX_EVAL="CC=clang && CXX=clang++ && PYTHON_VERSION=3.8.1 && PYTHON=python"
+  - name: "Ubuntu 18.04 + Python 3.6"
+    if: commit_message =~ /^\[release\]/
```
Comment (archibate) [inline .travis.yml]: Found Travis CI superior to AppVeyor!

**Comment by archibate** [comment]
https://github.com/taichi-dev/taichi/blob/6fcba966d75f5e665590b3a4c9cd5a9fa9e17851/.travis/install.sh#L14

What should we do when `$TRAVIS_OS_NAME == 'linux'`? I guess just replace `brew` by `apt-get` would work?

**Comment by yuanming-hu** [comment]
> https://github.com/taichi-dev/taichi/blob/6fcba966d75f5e665590b3a4c9cd5a9fa9e17851/.travis/install.sh#L4-L15
> 
> What should we do when `$TRAVIS_OS_NAME == 'linux'`? I guess just replace `brew` by `apt-get` would work?

I guess a better reference is the Jenkinsfile:
https://github.com/taichi-dev/taichi/blob/6fcba966d75f5e665590b3a4c9cd5a9fa9e17851/Jenkinsfile#L120

**Comment by archibate** [comment]
> https://github.com/taichi-dev/taichi/blob/6fcba966d75f5e665590b3a4c9cd5a9fa9e17851/Jenkinsfile#L23

Do we have `python3.6` in Travis?

**Comment by archibate** [comment]
Why don't Travis restart??

**Comment by archibate** [comment]
Fake release success! Six jobs running! See https://github.com/taichi-dev/taichi/pull/653/checks?check_run_id=532477013

File: .travis.yml (hunk 1 above)
Comment (yuanming-hu) [reply .travis.yml]: Cool! Much easier than AppVeyor!

**Comment by yuanming-hu** [comment]
Awesome!

**Comment by archibate** [comment]
wget `taichi-llvm.zip` doesn't work well, why not add `taichi-{llvm,clang}.zip` as a submodule of taichi? Namely `taichi-deps` maybe.

**Comment by yuanming-hu** [comment]
> wget `taichi-llvm.zip` doesn't work well, why not add `taichi-{llvm,clang}.zip` as a submodule of taichi? Namely `taichi-deps` maybe.

Oh, `taichi-llvm.zip` only has pre-built binaries for Windows. We should generate something like `taichi-llvm-8.0.1-linux-x64.zip`...

And you are right that downloading that file sometimes gets 404, but the probability is small. submodules are usually used to hold source files instead of binaries, so maybe it's better to stick to the current way?

**Comment by yuanming-hu** [comment]
I'm creating and uploading `taichi-llvm-8.0.1-linux-x64.zip`...

**Comment by yuanming-hu** [comment]
Uploaded. See here: https://github.com/yuanming-hu/taichi_assets/releases/tag/llvm8

File: .travis.yml [hunk 2]
```diff
@@ -30,15 +52,15 @@ cache:
...
-- echo $PYTHON
-- sysctl -n hw.ncpu
+- echo $TRAVIS_OS_NAME
 - pwd
+- echo $TRAVIS_BUILD_DIR
 - cd $TRAVIS_BUILD_DIR
 - mkdir taichi-llvm
 - cd taichi-llvm
-- wget https://github.com/yuanming-hu/taichi_assets/releases/download/llvm8/taichi-llvm-8.0.1.zip --retry-on-http-error=403 --waitretry=3 --tries=5
+- wget https://github.com/yuanming-hu/taichi_assets/releases/download/llvm8/taichi-llvm-8.0.1.zip --retry-on-http-error=403 --waitretry=3 --tries=5 || wget https://github.com/yuanming-hu/taichi_assets/releases/download/llvm8/taichi-llvm-8.0.1.zip
```
Comment (yuanming-hu) [inline .travis.yml]: https://github.com/yuanming-hu/taichi_assets/releases/download/llvm8/taichi-llvm-8.0.1.zip is for OS X.
We should use 
https://github.com/yuanming-hu/taichi_assets/releases/download/llvm8/taichi-llvm-8.0.1-linux-x64.zip for Linux.

Comment (yuanming-hu) [reply .travis.yml]: This error is due to running an OS X executable on Linux: https://travis-ci.com/github/taichi-dev/taichi/jobs/302348630#L1817

You can fix that by using the Linux `taichi-llvm-XXX`.zip

**Comment by yuanming-hu** [comment]
It may also make sense to `apt-get clang++-7` on Ubuntu 18.04 builds, since we need that to compile the runtime LLVM bitcode.

**Comment by archibate** [comment]
> It may also make sense to `apt-get clang++-7` on Ubuntu 18.04 builds, since we need that to compile the runtime LLVM bitcode.

Maybe we can specify CLANG_VERSION=7 in MATRIX_EVAL?

**Comment by yuanming-hu** [comment]
> > It may also make sense to `apt-get clang++-7` on Ubuntu 18.04 builds, since we need that to compile the runtime LLVM bitcode.
> 
> Maybe we can specify CLANG_VERSION=7 in MATRIX_EVAL? And test against 7,8,9 in [release].

Any version is fine. Note that we need clang only in the build process. Once the bitcode is built, we'll just ship it, and the static-linked LLVM in taichi will take over. So we don't have to test against 7, 8, 9.

**Comment by archibate** [comment]
We only found clang-8 for Ubuntu: https://pkgs.org/search/?q=clang, hope that's ok.

**Comment by archibate** [comment]
Do we need libclang?

**Comment by yuanming-hu** [comment]
clang-8 is fine and we probably don't need libclang. Thanks!

**Comment by yuanming-hu** [comment]
According to https://travis-ci.com/github/taichi-dev/taichi/jobs/303324626#L196 it seems that we are still using ubuntu 16.04. I'm not sure which version of clang is shipped with that ubuntu distribution.

**Comment by archibate** [comment]
```
/home/travis/build/taichi-dev/taichi/taichi/program/program.h:5:10: fatal error: 
      'optional' file not found
#include <optional>
```

**Comment by yuanming-hu** [comment]
According to https://stackoverflow.com/questions/47396959/does-clang6-implement-stdoptional it seems that we need gcc-7 which comes with C++17 headers. I guess an easier way is to use Ubuntu 18.04, and then you can `sudo apt install gcc-7`: https://docs.travis-ci.com/user/reference/bionic/

**Comment by archibate** [comment]
Note: `[dev]` means developer tools, CI is one of them.

**Comment by archibate** [comment]
e73c127 passing, great!

**Comment by archibate** [comment]
NEXT: how to install glew/glfw?

**Comment by archibate** [comment]
> [e73c127](https://github.com/taichi-dev/taichi/commit/e73c127ef763c52f865c34c99d58759fc10ca09d) passing, great!

NOT AT ALL: See https://travis-ci.com/github/taichi-dev/taichi/jobs/304809359, still many test failed, and `python -m taichi` doesn't provides a return value..

File: taichi/codegen/codegen_opengl.cpp [hunk 3]
```diff
@@ -381,9 +381,39 @@ int _rand_i32()\n\
       parent_type = root_snode_type_name_;
     }
 
-    emit("{}_ch {} = {}_children({}, {});", stmt->snode->node_type_name,
-         stmt->raw_name(), parent_type, parent->raw_name(),
-         stmt->input_index->raw_name());
+    const auto *sn = stmt->snode;
```
Comment (yuanming-hu) [inline taichi/codegen/codegen_opengl.cpp]: Could you please remove this changeset? It seems unrelated to this PR. Thanks!

File: .travis/install.sh [hunk 4]
```diff
@@ -12,4 +12,15 @@ if [[ $TRAVIS_OS_NAME == 'osx' ]]; then
     pyenv install $PYTHON_VERSION
     pyenv global $PYTHON_VERSION
     pyenv rehash
+elif [[ $TRAVIS_OS_NAME == 'linux' ]]; then
+    export PATH="$HOME/.pyenv/bin:$PATH"
+    eval "${MATRIX_EVAL}"
+    echo "python version: $PYTHON_VERSION"
+    #sudo apt-get update > /dev/null
```
Comment (yuanming-hu) [inline .travis/install.sh]: We can remove the comments here.

**Comment by archibate** [comment]
Don't merge, still have a major problem: `python -m taichi` didn't return non-zero even if test fails.

**Comment by yuanming-hu** [comment]
> > [e73c127](https://github.com/taichi-dev/taichi/commit/e73c127ef763c52f865c34c99d58759fc10ca09d) passing, great!
> 
> NOT AT ALL: See https://travis-ci.com/github/taichi-dev/taichi/jobs/304809359, still many test failed, and `python -m taichi` doesn't provides a return value..

Sounds like clang/llvm version issue. See 
https://github.com/pytorch/glow/issues/2442
https://github.com/crystal-lang/crystal/issues/6754

**Comment by yuanming-hu** [comment]
I have to sleep now. Almost 2am here... See you tomorrow. Thanks for your hard work!

**Comment by archibate** [comment]
Try fixing...if 81526f8 fails, we success. Have a good dream!

**Comment by archibate** [comment]
```
tests/python/test_ad_atomic.py::test_ad_reduce llvm-as: assembly parsed, but does not verify as correct!

Intrinsic has incorrect argument type!

void (i8*, i8*, i64, i1)* @llvm.memcpy.p0i8.p0i8.i64

Intrinsic has incorrect argument type!

void (i8*, i8*, i64, i1)* @llvm.memcpy.p0i8.p0i8.i64

Intrinsic has incorrect argument type!

void (i8*, i8*, i64, i1)* @llvm.memcpy.p0i8.p0i8.i64

Intrinsic has incorrect argument type!

void (i8*, i8*, i64, i1)* @llvm.memcpy.p0i8.p0i8.i64

Intrinsic has incorrect argument type!

void (i8*, i8*, i64, i1)* @llvm.memcpy.p0i8.p0i8.i64

Intrinsic has incorrect argument type!

void (i8*, i8*, i64, i1)* @llvm.memcpy.p0i8.p0i8.i64

Intrinsic has incorrect argument type!

void (i8*, i8, i64, i1)* @llvm.memset.p0i8.i64

Intrinsic has incorrect argument type!

void (i8*, i8, i64, i1)* @llvm.memset.p0i8.i64

Intrinsic has incorrect argument type!
```
Any idea?

**Comment by archibate** [comment]
Bad news, ubuntu bionic doesn't have 3.7 in pyenv.

**Comment by archibate** [comment]
Switched to 3.6, passed, seems the problem.

**Comment by yuanming-hu** [comment]
> ```
> tests/python/test_ad_atomic.py::test_ad_reduce llvm-as: assembly parsed, but does not verify as correct!
> 
> Intrinsic has incorrect argument type!
> 
> void (i8*, i8*, i64, i1)* @llvm.memcpy.p0i8.p0i8.i64
> 
> Intrinsic has incorrect argument type!
> 
> void (i8*, i8*, i64, i1)* @llvm.memcpy.p0i8.p0i8.i64
> 
> Intrinsic has incorrect argument type!
> 
> void (i8*, i8*, i64, i1)* @llvm.memcpy.p0i8.p0i8.i64
> 
> Intrinsic has incorrect argument type!
> 
> void (i8*, i8*, i64, i1)* @llvm.memcpy.p0i8.p0i8.i64
> 
> Intrinsic has incorrect argument type!
> 
> void (i8*, i8*, i64, i1)* @llvm.memcpy.p0i8.p0i8.i64
> 
> Intrinsic has incorrect argument type!
> 
> void (i8*, i8*, i64, i1)* @llvm.memcpy.p0i8.p0i8.i64
> 
> Intrinsic has incorrect argument type!
> 
> void (i8*, i8, i64, i1)* @llvm.memset.p0i8.i64
> 
> Intrinsic has incorrect argument type!
> 
> void (i8*, i8, i64, i1)* @llvm.memset.p0i8.i64
> 
> Intrinsic has incorrect argument type!
> ```
> 
> Any idea?

How did you fix this? It seems to me that everything passed...

**Comment by yuanming-hu** [comment]
> Switched to 3.6, passed, seems the problem.

I think we can simply `apt install python3.7-dev` and follow this link for 3.8:https://linuxize.com/post/how-to-install-python-3-8-on-ubuntu-18-04/

File: .travis/get-taichi-deps.sh [hunk 5]
```diff
@@ -7,7 +7,7 @@ if [ $TRAVIS_OS_NAME == 'osx' ]; then
 	wget https://github.com/yuanming-hu/taichi_assets/releases/download/llvm8/taichi-llvm-8.0.1.zip --retry-on-http-error=403 --waitretry=3 --tries=5
 	unzip taichi-llvm-8.0.1.zip
 elif [ $TRAVIS_OS_NAME == 'linux' ]; then
-	wget https://github.com/yuanming-hu/taichi_assets/releases/download/llvm8/taichi-llvm-8.0.1-linux-x64.zip
+	wget https://github.com/yuanming-hu/taichi_assets/releases/download/llvm8/taichi-llvm-8.0.1-linux-x64.zip --retry-on-http-error=403 --waitretry=3 --tries=5
```
Comment (archibate) [inline .travis/get-taichi-deps.sh]: No.. `wget` of Ubuntu seems can't understand these arguments... That's why I removed them.

Comment (yuanming-hu) [reply .travis/get-taichi-deps.sh]: I see. On my Ubuntu 18.04 machine, it actually works. Maybe we are talking about different versions of `wget`?

Comment (archibate) [reply .travis/get-taichi-deps.sh]: Maybe Travis used a minimized version of `wget`...从python默认2.7，和pyenv最高支持到3.6这两点就可以看出。

**Comment by archibate** [comment]
Found reason:
> If you decide to use apt anyway, note that for compatibility reasons, you’ll only be able to use the default Python versions that are available in Ubuntu (e.g. for Xenial, this means 2.7.12 and 3.5.1).

**Comment by archibate** [comment]
I really wish I could have a SSH tunnel to CI to check how it's `pyenv` works...

**Comment by yuanming-hu** [comment]
> I really wish I could have a SSH tunnel to CI to check how it's `pyenv` works...

On travis if it's too hard to make `pyenv` work, maybe we should simply `apt install python3.x-dev` instead, since Travis already provides an isolated *environment*/*container*.

**Comment by archibate** [comment]
Still `python` stands for `python2.7` now:
```
('Current version:', sys.version_info(major=2, minor=7, micro=17, releaselevel='final', serial=0))
```
